### PR TITLE
docs: document booting of ReaR rescue initramfs on z/VM

### PR DESCRIPTION
**Pull request by lzaoral**
##### Pull Request Details:

* Type: **Enhancement**

* Impact: **High**

* Reference to related issue (URL):
https://github.com/rear/rear/issues/3149#issuecomment-1943375764

* How was this pull request tested?

All steps described in the new documentation were tried on a Fedora Rawhide s390x VM running under the z/VM hypervisor.

* Description of the changes in this pull request:

Document booting of ReaR rescue initramfs on s390/s390x.

### CODE REVIEW

**Comment by jsmeix** [comment]
@lzaoral 
Wow!
Looks "rather interesting" to me as a total ZIPL noob ;-)
Now I have a starting point :-)
Thank you for it!

**Comment by jsmeix** [comment]
@lzaoral 
I would like to suggest to add a bit more context description
so that the "Booting ReaR rescue initramfs on s390/s390x"
section is somewhat self-contained (to a reasonable extent)
so that when someone reads only this section
it is comprehensible on its own.

Perhaps as a total ZIPL noob I am not an intended reader
of this section so my following suggestions could go too far.
Nevertheless I just make them and you can then decide
if and what context description you may add
as you think what is best for our users.

(1) Bare metal recovery:
As far as I see it seems the described method using zipl
makes the ReaR recovery system boot on the same system
where the zipl command is run.
Accordingly it seems the zipl method has the same drawback
as the kexec method, that a running Linux system is
already needed where the ReaR recovery system should boot
(to run kexec or zipl) so neither the kexec method
nor the zipl method support true bare metal recovery.
Is this the case or do I misunderstand something?
If my understanding is right, it should be clearly described
that a running Linux system is already needed
where the ReaR recovery system should be booted.
If I misunderstand things, it should be clearly described
where "rear mkbackup" was run
versus where "rear recover" should be run
and how to make the ReaR recovery system boot
where "rear recover" should be run.

(2) REAR_OUTPUT usage:
It confuses me that REAR_OUTPUT is used
both for zipl "output" via --target and
for zipl "input" via --image and --ramdisk.
I read
https://manpages.debian.org/jessie/s390-tools/zipl.8
and I am wondering if it is really required that one same
REAR_OUTPUT directory must be used for the zipl options
```
--target=REAR_OUTPUT
--image=REAR_OUTPUT/...
--ramdisk=REAR_OUTPUT/...
```
or if all could be different directories like
```
--target=BOOTMAP_INSTALL_DIRECTORY
--image=/SOME/PATH/TO/KERNEL
--ramdisk=/ANOTHER/PATH/TO/INITRD
```
If all could be different directories it should be clearly
described what the reason is why it is usually OK
to use one same directory for all three options.
I think the same directory for KERNEL and INITRD is natural
but perhaps a differnt --target directory could be required,
see the next item (3):

(3) Bootloader install device:
It was inexplicable to me why
```
The bootloader ... was installed on the dasda device.
```
regardless that no bootloader install device was specified.
I think
https://manpages.debian.org/jessie/s390-tools/zipl.8
explains it:
```
-t <TARGET DIRECTORY> or --target=<TARGET DIRECTORY>
Use the specified <TARGET DIRECTORY>.
zipl uses this directory to store the bootmap,
i.e. a file containing boot data.
The actual boot loader is installed onto the device
containing the target directory.
Supported devices are DASD and SCSI disks.
```
So it should be made more clear that via --target
the bootloader install device is implicitly specified
to avoid possibly "surprising" results when a user
specifies some --target=<TARGET DIRECTORY> that
happens to be on whatever unexpected device
where then a bootloader gets installed.
In particular when REAR_OUTPUT is used for --target
(i.e. the ReaR recovery system kernel may be stored
anywhere where its underlying device is not meant
to be booted from).

(4) Only for DASD disks:
According to
https://manpages.debian.org/jessie/s390-tools/zipl.8
the zipl --target method is only supported for
DASD and SCSI disks.
If this is right, it may be clearly described.
Probably - as far as I remember the code -
currently only DASD disks are supported by ReaR
so currently the zipl --target method is supported on
all IBM Z disks that are currently supported by ReaR.
The crucial wird is "currently" - i.e. if ReaR gets
enhanced to also support other IBM Z disks, the
described zipl --target method may no longer work
so to be future-proof this restriction of the
zipl --target method should be mentioned
because in practice nobody updates documentation ;-)

**Comment by jsmeix** [comment]
To avoid misunderstanding:

In my above
https://github.com/rear/rear/pull/3158#issuecomment-1956565746
I wrote about "bare metal recovery",
I even wrote "true bare metal recovery".

This was very misleading because on IBM Z systems
"true bare metal recovery" is not wanted.
On IBM Z there is always some virtualization in between.
What I actually meant was that the ReaR recovery system
can be booted on a "bare VM" i.e. on a VM where not yet
another Linux operating system must be running.

**Comment by lzaoral** [comment]
Thank you a lot for the thorough review, @jsmeix!

I've specified that the new section is only for virtual machines running under z/VM.  While it should be possible to also use ReaR on IBM Z LPARs, I have never tried that personally (maybe @pcahyna has some experience?).

(1) The `zipl` method can be used independently of the running OS which is an information I forgot include by using z/VM Control Program directly.  Actually, it should be quite straightforward to use `zipl` to implement `OUTPUT=USB` just like you would use GRUB2 or EXTLINUX to prepare a bootable device on `x86_64`.

(2) Correct, the value of the `--target=` option can be arbitrary as long as that directory is on the same device.  (Actually, I was able to create the bootmap file (even without `--force`) on a completely different device but unsurprisingly, the machine then failed to boot.)

(3) Hopefully done.

(4) RHEL only supports IBM Z machines with DASD or SCSI disks so I have zero experience with other disk types.  However, `zipl(8)` on Fedora Rawhide also mentions NVMe as a supported `--target=`.  Also, it seems that `--tape=` is the right option for tape devices.  So the only option that cannot be covered by `zipl` is the z/VM reader (a virtual punch card reader) which cannot be used with plain kernel and initramfs anyway (I have never tried booting RHEL installer from this device but, apparently, you also need some additional files to do so.).

I've updated the `zipl` section to include this information as well.

File: doc/user-guide/04-scenarios.adoc [hunk 1]
```diff
@@ -67,6 +67,103 @@ a single confidential command to let the redirection also apply for 'set -x'.
...
+
+1. Using the `kexec` tool as described in link:../../usr/share/rear/conf/templates/RESULT_usage_RAMDISK.txt[RESULT_usage_RAMDISK.txt],
+   if you have access to a running Linux instance.
+
+2. Using the `zipl` tool for management of boot devices on s390/s390x.
+
+=== Booting ReaR rescue initramfs on IBM Z (s390x) under z/VM with `zipl`
+
+After creating the rescue image with `rear mkrescue`, move the kernel and
+initramfs to the target directory (denoted by `REAR_DIRECTORY` in this example)
```
Comment (jsmeix) [inline doc/user-guide/04-scenarios.adoc:83]: I think using the term 'target directory' here
it is still misleading/confusing because
REAR_DIRECTORY is the "source" directory
where zipl gets kernel and initrd from
while the target is BOOTMAP_INSTALL_DIRECTORY

File: doc/user-guide/04-scenarios.adoc [hunk 2]
```diff
@@ -67,6 +67,103 @@ a single confidential command to let the redirection also apply for 'set -x'.
...
+    parameters......: 0x00009000-0x00009fff
+    kernel image....: 0x00010000-0x00cdbfff
+    parmline........: 0x00cdc000-0x00cdcfff
+    initial ramdisk.: 0x00cf0000-0x10980fff
+    environment blk.: 0x00cec000-0x00cecfff
+Preparing boot device for CCW- and LD-IPL: dasda.
+Done.
+```
+You can set `BOOTMAP_INSTALL_DIRECTORY` to any directory on the same device
+with `REAR_DIRECTORY` to store the boot data. The bootloader in the example was
```
Comment (jsmeix) [inline doc/user-guide/04-scenarios.adoc:119]: Sorry if I am rather nitpicking here
but I still do not fully understand things.

Here it looks as if BOOTMAP_INSTALL_DIRECTORY
is restricted by REAR_DIRECTORY so if for example
REAR_DIRECTORY is on a device that is not meant
to be booted from then the user may falsely set
BOOTMAP_INSTALL_DIRECTORY to a directory on that device.

I think the restiriction (if it exisist at all, see below)
is the other way round:
If /dev/dasda should be used to boot from
then BOOTMAP_INSTALL_DIRECTORY must be on /dev/dasda
and also REAR_DIRECTORY must be on /dev/dasda.

But perhaps only BOOTMAP_INSTALL_DIRECTORY
must be on /dev/dasda while REAR_DIRECTORY could
be on any device?

I fail to see in
https://manpages.debian.org/jessie/s390-tools/zipl.8
if REAR_DIRECTORY must be on the same disk device
where BOOTMAP_INSTALL_DIRECTORY is.

Simply put:
I do not understand if REAR_DIRECTORY is only a
"source" directory where zipl takes kernel and initrd from
and copies them (always of perhaps only if needed)
to the target destination on the boot disk that is
implicitly specified by BOOTMAP_INSTALL_DIRECTORY
OR
if kernel and initrd stay in REAR_DIRECTORY so that
REAR_DIRECTORY must be also on the boot disk that is
implicitly specified by BOOTMAP_INSTALL_DIRECTORY ?

**Comment by jsmeix** [comment]
@lzaoral 
some years ago (at the time when I did my first tests
with ReaR on IBM Z) we had an internal issue at SUSE
whether or not to officially support ReaR on IBM Z.
At that time only virtual machines running under z/VM
with only DASD disks were considered.
ReaR on IBM Z LPARs was explicitly not considered.
So at least for now we do not need to worry about
support for ReaR on IBM Z LPARs.
Nevertheless it is very good to metion that
support for ReaR on IBM Z is currently limited to
virtual machines running under z/VM with DASD disks.

File: doc/user-guide/04-scenarios.adoc (hunk 2 above)
Comment (lzaoral) [reply doc/user-guide/04-scenarios.adoc]: > kernel and initrd stay in REAR_DIRECTORY so that
REAR_DIRECTORY must be also on the boot disk that is
implicitly specified by BOOTMAP_INSTALL_DIRECTORY

This is what happens.  (The manual page does not say that but as noted in (2) https://github.com/rear/rear/pull/3158#issuecomment-1964207577, `zipl(8)` won't necessarily stop you to use different devices, but the machine will not boot!)

I've reworded the section to make it more explicit.

**Comment by jsmeix** [comment]
Now this
```
The machine will not boot if `BOOTMAP_INSTALL_DIRECTORY`
and `REAR_DIRECTORY` are on different devices!
```
makes things clear and it explains why in practice
it is probably simplest and best to use one same
directory for bootmap, kernel, and initrd.

**Review by jsmeix — APPROVED** [review]
Thank you for such valuable documentation!

**Comment by jsmeix** [comment]
@pcahyna 
because in
https://github.com/rear/rear/pull/3158#issuecomment-1964207577
@lzaoral "advertised" you as a possible "IBM Z expert" ;-)
I would appreciate it if you colud have a look here
(as time permits).

@rear/contributors 
I would like to merge it on Friday afternoon
unless there are objections.

**Comment by pcahyna** [comment]
@jsmeix sorry for being late, looking now ...

**Comment by pcahyna** [comment]
can you please delay merging a bit?

**Comment by jsmeix** [comment]
@pcahyna 
don't worry!
Of course I delay merging when you need time for review.
No rush, take your time, and
have a relaxed and recovering weekend!

File: doc/user-guide/04-scenarios.adoc [hunk 3]
```diff
@@ -67,6 +67,106 @@ a single confidential command to let the redirection also apply for 'set -x'.
...
+Loadparm:    ""
+Bootparms:   ""
+clear:       0
+# reboot
+...
+```
+
+If you do not have a running Linux instance at hand, you can force the boot
+from the given device using the z/VM Control Program (CP) directly, e.g. using
+a serial console.
```
Comment (pcahyna) [inline doc/user-guide/04-scenarios.adoc:142]: I don't think there is a serial console on the S/390 - we are not in Unix world like on the PowerPC/PReP or Sparc machines. There is 3270 which AFAIK is a very different beast. There is also 3215 which may be sufficiently similar to a serial connection but I don't think it is a serial connection really, not sure.

File: doc/user-guide/04-scenarios.adoc [hunk 4]
```diff
@@ -67,6 +67,106 @@ a single confidential command to let the redirection also apply for 'set -x'.
 See the description in /usr/share/rear/conf/default.conf
 how to set variables for secret values in a confidential way.
 
+== Booting ReaR rescue initramfs on IBM Z (s390x) under z/VM
+
+There are two options how to boot the rescue environment on IBM Z (s390x) under
+z/VM created using the `OUTPUT=RAMDISK` setting.
```
Comment (pcahyna) [inline doc/user-guide/04-scenarios.adoc:73]: Other scenarios in this document use a config file snippet. Like
> If you want a bootable USB device with a (tar) backup to USB as well, you would use:
> ```
> BACKUP=NETFS
> OUTPUT=USB
> USB_DEVICE=/dev/disk/by-label/REAR-000
> ```

I think you should add this kind of snippet as well.

File: doc/user-guide/04-scenarios.adoc [hunk 5]
```diff
@@ -67,6 +67,106 @@ a single confidential command to let the redirection also apply for 'set -x'.
...
+1. Using the `kexec` tool as described in link:../../usr/share/rear/conf/templates/RESULT_usage_RAMDISK.txt[RESULT_usage_RAMDISK.txt],
+   if you have access to a running Linux instance.
+
+2. Using the `zipl` tool for management of boot devices on s390/s390x.
+
+=== Booting ReaR rescue initramfs on IBM Z (s390x) under z/VM with `zipl`
+
+After creating the rescue image with `rear mkrescue`, move the kernel and
+initramfs to a directory (denoted by `REAR_DIRECTORY` in this example) on the
+target device to boot from.
```
Comment (pcahyna) [inline doc/user-guide/04-scenarios.adoc:84]: I know that this has been discussed between you and @jsmeix a lot, but I think the use of `REAR_DIRECTORY` (and also `BOOTMAP_INSTALL_DIRECTORY`) is confusing. Since you are giving an example and in real use one will most likely use an absolute path as the argument to `zipl`, I think it is better to provide a realistic example, rather than providing a placeholder that the user should replace. Also, I think you should then describe more clearly what is the relation of the directory to the disk. You say that the directory is on the target device to boot from, which is correct, but maybe it should be written in a more explicit way (the style of this document is the one of specific examples, not the one of a reference manual), like this:
"Mount a partition on the disk that you want to make bootable. Let's say that the disk is called `/dev/dasda`, the partition with the filesystem that will contain the kernel and initramfs is `/dev/dasda1` and we mount it on `/mnt/bootdisk` :
```
# mount /dev/dasda1 /mnt/bootdisk
```
Copy the kernel and initramfs there:
```
# cp <something, TBD> /mnt/bootdisk
```
"
and from this point use `/mnt/bootdisk` instead of `REAR_DIRECTORY` and `BOOTMAP_INSTALL_DIRECTORY` in the example.
WDYT?

File: doc/user-guide/04-scenarios.adoc [hunk 6]
```diff
@@ -67,6 +67,106 @@ a single confidential command to let the redirection also apply for 'set -x'.
...
+    internal loader.: 0x0000a000-0x0000dfff
+    parameters......: 0x00009000-0x00009fff
+    kernel image....: 0x00010000-0x00cdbfff
+    parmline........: 0x00cdc000-0x00cdcfff
+    initial ramdisk.: 0x00cf0000-0x10980fff
+    environment blk.: 0x00cec000-0x00cecfff
+Preparing boot device for CCW- and LD-IPL: dasda.
+Done.
+```
+You must set `BOOTMAP_INSTALL_DIRECTORY` to any directory on the same device
```
Comment (pcahyna) [inline doc/user-guide/04-scenarios.adoc:118]: You don't set `BOOTMAP_INSTALL_DIRECTORY` to anything as it is not a variable. You could write "replace `BOOTMAP_INSTALL_DIRECTORY` by (...)", or keep it and replace `BOOTMAP_INSTALL_DIRECTORY` in your examples by `$BOOTMAP_INSTALL_DIRECTORY`, but I believe it is better to use a real directory here. See above.

File: doc/user-guide/04-scenarios.adoc [hunk 7]
```diff
@@ -67,6 +67,106 @@ a single confidential command to let the redirection also apply for 'set -x'.
...
+z/VM created using the `OUTPUT=RAMDISK` setting.
+
+1. Using the `kexec` tool as described in link:../../usr/share/rear/conf/templates/RESULT_usage_RAMDISK.txt[RESULT_usage_RAMDISK.txt],
+   if you have access to a running Linux instance.
+
+2. Using the `zipl` tool for management of boot devices on s390/s390x.
+
+=== Booting ReaR rescue initramfs on IBM Z (s390x) under z/VM with `zipl`
+
+After creating the rescue image with `rear mkrescue`, move the kernel and
```
Comment (pcahyna) [inline doc/user-guide/04-scenarios.adoc:82]: What is a "rescue image" when `OUTPUT=RAMDISK`? Do you call the kernel and initramfs a rescue image? And where do you find the kernel and initramfs to move?

File: doc/user-guide/04-scenarios.adoc (hunk 5 above)
Comment (jsmeix) [reply doc/user-guide/04-scenarios.adoc:84]: This part seems to be endlessly complicated
and all descriptions look confusing
as long as one does not understand
the IPL implementation details behind :-(

Now I am wondering how the IBM Z "initial bootloader"
(I don't know how that thingy is called in IBM Z speech,
 I mean what loads kernel and initrd and executes it)
accesses kernel and initrd on the boot disk.

With
```
zipl --target=BOOTMAP_INSTALL_DIRECTORY \
       --parameters='root=/dev/ram0 ro' \
       --image=REAR_DIRECTORY/kernel-HOSTNAME \
       --ramdisk=REAR_DIRECTORY/initramfs-HOSTNAME.img
```
it looks as if the "initial bootloader"
loads kernel and initrd via their provided
filesystem names i.e.
```
REAR_DIRECTORY/kernel-HOSTNAME
REAR_DIRECTORY/initramfs-HOSTNAME.img
```
but then
```
/mnt/bootdisk/kernel-HOSTNAME
/mnt/bootdisk/initramfs-HOSTNAME.img
```
would not exist at IPL time because
/mnt/bootdisk/ does not exist at IPL time
instead /kernel-HOSTNAME and /initramfs-HOSTNAME.img
exist in the filesystem on /dev/dasda1 at IPL time.

When
```
zipl --target=/mnt/bootdisk \
       --parameters='root=/dev/ram0 ro' \
       --image=/mnt/bootdisk/kernel-HOSTNAME \
       --ramdisk=/mnt/bootdisk/initramfs-HOSTNAME.img
```
works, it means the "initial bootloader"
accesses kernel and initrd not via the
filesystem on /dev/dasda1
but differently.

I guess the "initial bootloader"
accesses kernel and initrd as LILO did it
by accessing directly the hardware "sectors"
on /dev/dasda1 where kernel and initrd are stored.

When my guess is right, it also means that
each time when ReaR creates a new recovery system
(i.e. after each "rear mkrescue/mkbackup")
```
mount /dev/dasda1 /mnt/bootdisk

cp <something, TBD> /mnt/bootdisk

zipl --target=/mnt/bootdisk \
       --parameters='root=/dev/ram0 ro' \
       --image=/mnt/bootdisk/kernel-HOSTNAME \
       --ramdisk=/mnt/bootdisk/initramfs-HOSTNAME.img

umount /mnt/bootdisk
```
must be called.

I guess this is what
rear/finalize/Fedora/s390/660_install_zipl.sh
and
rear/finalize/SUSE_LINUX/s390/660_install_grub2_and_zipl.sh
implement (or try to implement).

Am I right or do I misunderstand things?

https://manpages.debian.org/jessie/s390-tools/zipl.8
is rather vague what the purpose of the 'bootmap' file is
but I guess in particular the hardware "sectors"
of kernel and initrd are stored in the 'bootmap' file.

By the way:
Right now I found in
https://manpages.debian.org/jessie/s390-tools/zipl.8
```
-a or --add-files
Copy all specified files to the bootmap file
instead of just referencing them.
This option allows specifying files in a boot configuration
which are not located on the target device.
```
This looks as if 'zipl -a' is a fail-safe method
to be able to still boot regardless what happened later
after the 'zipl' call with the kernel and initrd files
at their location in the filesystem on /dev/dasda1
i.e. changing /mnt/bootdisk/kernel-HOSTNAME
or /mnt/bootdisk/initramfs-HOSTNAME.img
won't destroy a boot mechanism with that files.

**Comment by jsmeix** [comment]
@pcahyna 
thank you for your review!
As always your precise view reveals problematic things.

File: doc/user-guide/04-scenarios.adoc (hunk 5 above)
Comment (pcahyna) [reply doc/user-guide/04-scenarios.adoc:84]: I admit that I also don't know if zipl is more like 'lilo' or more like 'grub' and this difference indeed matters from the POV of an administrator.

Comment (lzaoral) [reply doc/user-guide/04-scenarios.adoc]: I've made the steps more explicit in 873884f.  Thank you both for the suggestions!

File: doc/user-guide/04-scenarios.adoc (hunk 4 above)
Comment (lzaoral) [reply doc/user-guide/04-scenarios.adoc:87]: Added in 873884f.

File: doc/user-guide/04-scenarios.adoc (hunk 6 above)
Comment (lzaoral) [reply doc/user-guide/04-scenarios.adoc]: The guide now uses explicit paths as suggested in https://github.com/rear/rear/pull/3158#discussion_r1509456636.

File: doc/user-guide/04-scenarios.adoc (hunk 3 above)
Comment (lzaoral) [reply doc/user-guide/04-scenarios.adoc]: Thank you!  Changed in https://github.com/rear/rear/commit/873884f3831bc1d4345ce247df2e3196bd15b1d6.

File: doc/user-guide/04-scenarios.adoc [hunk 8]
```diff
@@ -67,6 +67,122 @@ a single confidential command to let the redirection also apply for 'set -x'.
...
+
+First, we have to select a partition on the disk that we want to make bootable.
+Assume that the disk is called `/dev/dasda` and the partition with the filesystem
+that will contain the kernel and initramfs is `/dev/dasda1`.
+
+We mount this partition to `/mnt/bootdisk` and copy there the kernel and initramfs
+created in the previous section:
+```
+# mount /dev/dasda1 /mnt/bootdisk
+# cp $OUTPUT_URL/HOSTNAME/* /mnt/bootdisk/
```
Comment (pcahyna) [inline doc/user-guide/04-scenarios.adoc:104]: 1. we have not set `OUTPUT_URL` anywhere in the shell so this won't work and 2. if we set it in the shell just as shown for local.conf, it will fail, because cp will not like a file:// URL. Please change to `/path/to/initramfs`.

File: doc/user-guide/04-scenarios.adoc [hunk 9]
```diff
@@ -67,6 +67,122 @@ a single confidential command to let the redirection also apply for 'set -x'.
...
+```
+# zipl --target=/mnt/bootdisk \
+       --parameters='root=/dev/ram0 ro' \
+       --image=/mnt/bootdisk/kernel-HOSTNAME \
+       --ramdisk=/mnt/bootdisk/initramfs-HOSTNAME.img
+Run /lib/s390-tools//zipl_helper.device-mapper /mnt/bootdisk/
+Building bootmap in '/mnt/bootdisk/'
+Adding IPL section
+  initial ramdisk...: /mnt/bootdisk/initramfs-ibm-z-523.img
+  kernel image......: /mnt/bootdisk/kernel-ibm-z-523
```
Comment (pcahyna) [inline doc/user-guide/04-scenarios.adoc:128]: inconsistency in example - either use HOSTNAME everywhere, or ibm-z-523 everywhere

File: doc/user-guide/04-scenarios.adoc [hunk 10]
```diff
@@ -67,6 +67,122 @@ a single confidential command to let the redirection also apply for 'set -x'.
...
+If you do not have a running Linux instance at hand, you can force the boot
+from the given device using the z/VM Control Program (CP) directly, e.g. through
+a terminal connection.
+
+The following example only considers DASD disk devices. Consult the official
+z/VM CP Manual for booting from other devices (e.g. FCP attached SCSI) or for
+additional details in general.
+
+Use the following command to boot from a DASD boot device through z/VM CP.
+Do not forget to include the `#` prefix which marks commands for z/VM CP!
```
Comment (pcahyna) [inline doc/user-guide/04-scenarios.adoc:164]: well... if you are at the CP or CMS prompt, you don't use the `#` prefix, do you?

File: doc/user-guide/04-scenarios.adoc [hunk 11]
```diff
@@ -67,6 +67,122 @@ a single confidential command to let the redirection also apply for 'set -x'.
...
+```
+#CP IPL vdev
```
Comment (pcahyna) [inline doc/user-guide/04-scenarios.adoc:166]: use `0120` instead of `vdev`?

File: doc/user-guide/04-scenarios.adoc [hunk 12]
```diff
@@ -67,6 +67,122 @@ a single confidential command to let the redirection also apply for 'set -x'.
...
+           which contains the given `--target` directory, in our case
+           `/dev/dasda`!
+
+Use the following command to make the device mounted at `/mnt/bootdisk` bootable
+with `zipl`:
+```
+# zipl --target=/mnt/bootdisk \
+       --parameters='root=/dev/ram0 ro' \
+       --image=/mnt/bootdisk/kernel-HOSTNAME \
+       --ramdisk=/mnt/bootdisk/initramfs-HOSTNAME.img
```
Comment (jsmeix) [inline doc/user-guide/04-scenarios.adoc:123]: Only a cosmetic nitpick:
I think adding an artificial empty line after the command
(in particular here when it is a multi-line command)
aids readablility like
```
# long command \
  split over \
  several lines

several lines of
command output
```
to make it easier for the user to see
where the multi-line command ends
and where the command output begins.

File: doc/user-guide/04-scenarios.adoc [hunk 13]
```diff
@@ -67,6 +67,122 @@ a single confidential command to let the redirection also apply for 'set -x'.
...
+================================================================================
+0120      active    dasda     94:0    ECKD         4096   39128MB   10017000
+```
+or directly from z/VM CP with the following query:
+```
+#CP QUERY VIRTUAL DASD
+00:
+00: CP QUERY VIRTUAL DASD
+00: DASD 0120 ON DASD  2213 R/W 0X0120 SUBCHANNEL = 0004
+00: DASD 0121 ON DASD  201A R/W 0X0121 SUBCHANNEL = 0005
```
Comment (jsmeix) [inline doc/user-guide/04-scenarios.adoc:183]: I have an understanding question here:
From above I see that '0120' is "the right one"
but I do not understand how to see that in the
output of the `#CP QUERY VIRTUAL DASD` command,
i.e. why '0121' is not "the right one"
in particular compared to the output of
the `lsdasd -s` command that shows only one disk
so that this one disk must be "the right one".

File: doc/user-guide/04-scenarios.adoc [hunk 14]
```diff
@@ -67,7 +67,21 @@ a single confidential command to let the redirection also apply for 'set -x'.
...
+[source,bash]
+----
+OUTPUT=RAMDISK
+OUTPUT_URL=file:///path/to/initramfs
+BACKUP=NETFS
+BACKUP_URL=nfs://NFS.server.IP/path/to/backup
+----
+
+Make sure that the path in `OUTPUT_URL` exists and create the rescue initramfs
+and backup and using `rear mkrescue`.
```
[the remark this answers is not in the document]
Comment (lzaoral) [reply doc/user-guide/04-scenarios.adoc]: Thanks!  Addressed in 7f73704142be5e7a88f07bdb0e08fdf3bd7645e0.

File: doc/user-guide/04-scenarios.adoc (hunk 9 above)
Comment (lzaoral) [reply doc/user-guide/04-scenarios.adoc]: My bad.  Addressed in https://github.com/rear/rear/commit/7f73704142be5e7a88f07bdb0e08fdf3bd7645e0.

File: doc/user-guide/04-scenarios.adoc (hunk 12 above)
Comment (lzaoral) [reply doc/user-guide/04-scenarios.adoc:123]: Sure!  Addressed in https://github.com/rear/rear/commit/7f73704142be5e7a88f07bdb0e08fdf3bd7645e0.

File: doc/user-guide/04-scenarios.adoc (hunk 10 above)
Comment (lzaoral) [reply doc/user-guide/04-scenarios.adoc]: TIL!  Addressed in https://github.com/rear/rear/commit/7f73704142be5e7a88f07bdb0e08fdf3bd7645e0.

File: doc/user-guide/04-scenarios.adoc (hunk 11 above)
Comment (lzaoral) [reply doc/user-guide/04-scenarios.adoc]: Addressed and reworded in https://github.com/rear/rear/commit/7f73704142be5e7a88f07bdb0e08fdf3bd7645e0.

File: doc/user-guide/04-scenarios.adoc (hunk 13 above)
Comment (lzaoral) [reply doc/user-guide/04-scenarios.adoc:184]: Yes, I also believe that there is no reliable way to tell which device is the right from CP.  (On the other hand, nothing stops you to just try `IPL` on all of them one by one).

By the way, it is possible to explicitly ignore devices when Linux boots which is the reason why CP may list more DASD devices than `lsdasd`.  See:
* `cioignore(8)` -  https://www.ibm.com/docs/en/linux-on-z?topic=commands-cio-ignore
* `cio-ignore=` kernel cmdline parameter - https://www.ibm.com/docs/en/linux-on-systems?topic=parameters-cio-ignore

File: doc/user-guide/04-scenarios.adoc [hunk 15]
```diff
@@ -67,6 +67,123 @@ a single confidential command to let the redirection also apply for 'set -x'.
...
+[source,bash]
+----
+OUTPUT=RAMDISK
+OUTPUT_URL=file:///path/to/initramfs
+BACKUP=NETFS
+BACKUP_URL=nfs://NFS.server.IP/path/to/backup
+----
+
+Make sure that the path in `OUTPUT_URL` exists and create the rescue initramfs
+and backup using `rear mkrescue`.
```
Comment (jsmeix) [inline doc/user-guide/04-scenarios.adoc:84]: "rear mkrescue" does not cerate a backup
so use `rear mkbackup` in the described case here, i.e.
"to create the rescue initramfs with a (tar) backup".

Comment (lzaoral) [reply doc/user-guide/04-scenarios.adoc:84]: Another typo...  Thank you, @jsmeix!

Comment (lzaoral) [reply doc/user-guide/04-scenarios.adoc]: Addressed in 74cb29a7.

**Comment by jsmeix** [comment]
As far as I see now all looks well
so I would like to merge it tomorrow afternoon
unless objections appear.

@pcahyna 
please approve this pull request when it is OK for you.

File: doc/user-guide/04-scenarios.adoc (hunk 13 above)
Comment (pcahyna) [reply doc/user-guide/04-scenarios.adoc:184]: There are actually two ways of ignoring devices: one is to ignore them completely by the kernel (`cio_ignore` does that), and marking them offline, which will avoid creating /dev/dasd* device nodes for them and will show them only in the `lsdasd --offline` output.
CP and CMS will usually see more devices than Linux, because some of them are used by CMS and showing them as device nodes under Linux is usually not helpful.

File: doc/user-guide/04-scenarios.adoc [hunk 16]
```diff
@@ -67,6 +67,123 @@ a single confidential command to let the redirection also apply for 'set -x'.
...
+================================================================================
+0120      active    dasda     94:0    ECKD         4096   39128MB   10017000
+```
+or directly from z/VM CP with the following query:
+```
+CP QUERY VIRTUAL DASD
+00:
+00: CP QUERY VIRTUAL DASD
+00: DASD 0120 ON DASD  2213 R/W 0X0120 SUBCHANNEL = 0004
+00: DASD 0121 ON DASD  201A R/W 0X0121 SUBCHANNEL = 0005
```
Comment (pcahyna) [inline doc/user-guide/04-scenarios.adoc:184]: Not part of the documentation changes, but I think it could be useful to set the label of the dasd formatted by ReaR to e.g. `REAR` so that it is immediately visible in `CP`/`CMS` that it contains the ReaR backup. This would belong to a hypothetical `USB` output method for DASDs, specifically the `rear format` step.

**Comment by lzaoral** [comment]
I've rebased the branch against `master` and squashed all fixups.

**Review by pcahyna — APPROVED** [review]

**Comment by jsmeix** [comment]
@rear/contributors 
I would like to merge it today afternoon
unless there are objections.

**Comment by jsmeix** [comment]
@lzaoral 
thank you for all your work to get that done properly and
for your patience to answer all my understanding questions!

@pcahyna 
thank you for your careful review!
